core/modules/wpublish/ChunkListQservRequest.cc

```diff
@@ -118,24 +118,41 @@ void ChunkListQservRequest::onResponse(proto::FrameBufferView& view) {
         }
         LOGS(_log, LOG_LVL_DEBUG, context << "total chunks removed: " << numRemoved);
     }
+    if (nullptr != _onFinish) {
 
-    if (_onFinish) {
-        _onFinish(
-            ::translate(reply.status()),
-            reply.error(),
-            added,
-            removed);
+        // Clearing the stored callback after finishing the up-stream notification
+        // has two purposes:
+        //
+        // 1. it guaranties (exactly) one time notification
+        // 2. it breaks the up-stream dependency on a caller object if a shared
+        //    pointer to the object was mentioned as the lambda-function's closure
+
```

Comment:
this is a great comment and an important mechanism. Its seems like the pattern is repeated in many places, and `CallbackType _onFinish` is a member of more than 1 class. I wonder if there is an opportunity here to refactor the `onFinish` mechanism into a common base class, and/or move the clear-and-call behavior into a worker function?

Reply:
The notifications are guaranteed to be called in a thread-safe environment. So, simply re-initializing the data members with 'nullptr' should be enough to solve this potential problem (as we discussed it in a Slack channel). And this is what I'm going to do to all occurrences of this pattern for now.

And I have already thought about an idea of simplifying the code which is similar to your suggestion to push the repeating pattern upstream to a base class. This idea has a few problems (and that was a reason why I decided not to go for it, or, at least, to postpone its implementation until I'll have a better idea how this could be implemented). The first problem is that a signature of each callback function depends on a specific subclass. The base class, for obvious reasons, down't have this information. The second problem is that different subclasses may potentially (and in some cases - do) invoke callbacks with different sets (and type) of parameters. Moreover, the value of those parameters is determined by a subclass at a time a notification is about to be sent up-stream to a caller of a class (a job, a request, etc.). Therefore wrapping the subclass-specific notification into some wrapper object and passing it to a base class would be a bit problematic since it's impossible to do the early binding of parameters. Nevertheless, it still feels like this is "doable". I will think more about this tomorrow.